src/main/scala/io/picnicml/doddlemodel/cluster/DBSCAN.scala

```diff
@@ -0,0 +1,87 @@
+package io.picnicml.doddlemodel.cluster
+
+import breeze.linalg.functions.euclideanDistance
+import cats.syntax.option._
+import io.picnicml.doddlemodel.data.Features
+import io.picnicml.doddlemodel.typeclasses.Clusterer
+
+/** An immutable DBSCAN model.
+  *
+  * @param eps: the maximum distance between points in a group
+  * @param minSamples: the minimum number of point in a core group
+  *
+  * Examples:
+  * val model = DBSCAN()
+  * val model = DBSCAN(eps = 1.5)
+  * val model = DBSCAN(minSamples = 3)
+  * val model = DBSCAN(eps = 2.0, minSamples = 3)
+  */
+case class DBSCAN private(eps: Double, minSamples: Int, private val label: Option[Array[Int]])
+
+object DBSCAN {
+
+  val NOISE: Int = -1
+  val UNASSIGNED: Int = Int.MaxValue
+  val GROUPSTART: Int = 0
+
+  def apply(eps: Double = 1.0, minSamples: Int = 1): DBSCAN = {
+    require(eps > 0.0, "Maximum distance needs to be larger than 0")
+    require(minSamples > 0, "Minimum number of samples needs to be larger than 0")
+    DBSCAN(eps, minSamples, none)
+  }
+
+  implicit lazy val ev: Clusterer[DBSCAN] = new Clusterer[DBSCAN] {
+
+    override def isFitted(model: DBSCAN): Boolean = model.label.isDefined
+
+    override protected def labelSafe(model: DBSCAN): Array[Int] = model.label.get
+
+    override protected def copy(model: DBSCAN): DBSCAN =
+      model.copy()
+
+    override protected def copy(model: DBSCAN, label: Array[Int]): DBSCAN =
+      model.copy(label = label.some)
+
+    override protected def fitSafe(model: DBSCAN, x: Features): DBSCAN = {
+      val label = Array.fill[Int](x.rows)(UNASSIGNED)
+      var groupId = GROUPSTART
+      for (pointId <- 0 until x.rows if label(pointId) == UNASSIGNED) {
+        var groupQueue = findNeighbors(pointId, x, model.eps)
+        if (groupQueue.size + 1 < model.minSamples) {
+          label(pointId) = NOISE
+        } else {
+          label(pointId) = groupId
+          while (groupQueue.nonEmpty) {
```

Comment:
Unfortunately, there still seems to be a problem with the sample points I've mentioned before, this time in a slightly different setting. The setting:
```scala
val x = DenseMatrix(List(0.0, 12.0), List(0.0, 9.0), List(0.0, 6.0), List(0.0, 3.0), List(0.0, 0.0))
val model = DBSCAN(3.0, 3)
DBSCAN.ev.fitPredict(model, x)
```
The result that is obtained is `Array(0, 0, 0, 1, 1)`, whereas all the points should really belong to cluster `0` (checked with sklearn to be sure).

To my understanding, the problem seems to be that the propagation of labels should happen slightly different for _"immediate neighbors"_ and _neighbors-of-neighbors_, but in the code it is currently handled the same way.
What I mean by that: in the pseudocode in paper, 
1. the cluster ID for immediate neighbors (`seeds`) is always updated (assuming that there are enough immediate neighbors)
2. the cluster ID for neighbors-of-neighbors is updated only if the current immediate neighbor being processed has `>= minSamples` neighbors in its Eps-neighborhood and the neighbor-of-neighbor is either unassigned or a noise point.

The above paragraph tries to paraphrase the part of pseudocode in the paper after the  
`ELSE // all points in seeds are density-reachable from Point` part.

Please let me know if something is not clear (I'm a bit tired and I'm not sure if I'm making sense at the moment :)) and I'll try to explain it differently.

Reply:
Actually, the reason why your testing is failing is that the distance map is not cleared.
and the `>=` point that you stated.
Fixed the issue and added the test case.